Kafka headers support for HTTP

codeJack opened the pull request:
Purpose of the change is to allow usage of [Kafka Headers](https://issues.apache.org/jira/browse/KAFKA-4208) within strimzi-kafka-bridge HTTP API.
Kafka Headers would be attached to the [ConsumerRecord](https://strimzi.io/docs/bridge/latest/#_consumerrecord) and the [ProducerRecord](https://strimzi.io/docs/bridge/latest/#_producerrecord) definitions as an _optional_ array of key/value pairs.

Example

```
"kafkaRecord": {
  "key": "key1",
  "partition": 0,
  "value": "value1",
  "headers": [
    {
      "key": "key1",
      "value": "value1"
    },
    {
      "key": "key2",
      "value": "value2"
    }
  ]
}
```

 The feature is also tracked by this issue : https://github.com/strimzi/strimzi-kafka-bridge/issues/183

codeJack commented:
Looks like CI failed because of no space left on device
![image](https://user-images.githubusercontent.com/1332227/82817050-916abe00-9e9c-11ea-8765-dafea552c755.png)

scholzj commented:
I restarted the build -> maybe Travis cleaned the disks in the mean time. If not we will estart it later. I passed on amd64, so I think thsi should not indicate any issue with this PR.

codeJack commented:
Thanks @scholzj.
CI went a bit further, that said the failing test cases are not failing in local.
Could it be that they are unstable or do I have to look into the way I validated locally ?
Am I allowed to restart the build ?

scholzj commented:
TBH, I'm not sure how Travis decides who can and connot restart the builds ... seems a bit random sometimes :-). But I wouldn't worry about it for now.

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/http/converter/HttpJsonMessageConverter.java:
```diff
@@ -25,6 +28,7 @@
         Integer partitionFromBody = null;
         byte[] key = null;
         byte[] value = null;
+        MultiMap headers = MultiMap.caseInsensitiveMultiMap();
```
why not filling a `List<KafkaHeader>` that the Vert.x client provides and then adding the list to the record at the end instead of filling this map and then traversing it for adding one header at time?

codeJack replied:
Indeed, I have update the PR accordingly.

ppatierno commented on src/main/resources/openapi.json:
```diff
@@ -1104,13 +1104,60 @@
+                        "type": "string"
+                    },
+                    "value": {
+                        "type": "string"
```
I would like to start a discussion around this.
As for the Kafka protocol, the header value is just an array of bytes so it could contain binary data as well and not just string. Did you think about that? Is there any specific reason you decided to support string only? for your use case?
As for the value message body, the current bridge support JSON and binary encoding which is specified in the `format` field when a consumer is created. Should we provide support for a more generic possible way to encode header value as well? Just thinking aloud ..... @stanlyDoge @scholzj wdyt?

codeJack replied:
String was enough for my use case, but indeed it might not be generic enough.
I can add a `format` field (default should probably be` String` ?) at `KafkaHeader` level that would qualify the format of the header value.

```
"kafkaRecord": {
  "key": "key1",
  "partition": 0,
  "value": "value1",
  "headers": [
    {
      "key": "key1",
      "value": "string default format"
    },
    {
      "key": "key2",
      "format": "string"
      "value": "string format"
    },
    {
      "key": "key3",
      "format": "binary"
      "value": "<base 64 encoded bytes>"
    }
  ]
}
```

codeJack replied:
Yes, I guess I got your point  : sticking to Kafka's wire format for headers  - that would be raw binary - as  the official Kafka APIs do not provided facilities to serialize / deserialize Headers in the first place. 
In other words you'd like the Strimzi HTTP bridge to provide the same level of service as the official APIs for what concerns headers.

While this makes perfectly sense, in real life headers will be mostly used to express messaging metadata and I assume that this would be commonly done in the form of key/value String pairs; an hypothetical base64 encoding / decoding overhead is minimal but I guess it would be handy to benefit from human readable headers within the bridge interfaces to further ease investigations and testing.

ppatierno replied:
@codeJack now that we agreed on the base64 encoded value, we should document it.
I think that in the "value" and alongside "type", the OpenAPI spec should support a "description" to make it clear. Wdyt?

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/facades/KafkaFacade.java:
```diff
@@ -74,6 +93,20 @@ public void produce(String topic, byte[] bytes, int messageCount, int partition)
+        Headers kafkaHeaders = new RecordHeaders();
+        for (Map.Entry<String, String> header: headers) {
+            kafkaHeaders.add(new RecordHeader(header.getKey(), header.getValue().getBytes()));
+        }
```
maybe you can use `headers.forEach` ?

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/ConsumerTest.java:
```diff
@@ -565,6 +628,72 @@ void receiveBinaryMessage(VertxTestContext context) throws InterruptedException,
     }
 
+    @Test
+    void receiveBinaryMessageWithHeaders(VertxTestContext context) throws InterruptedException, ExecutionException, TimeoutException {
```
I don't know how much it makes sense to have a different test with binary value when it doesn't impact the headers. Wdyt @see-quick ?

codeJack replied:
Given that I updated both [HttpJsonMessageConverter](https://github.com/strimzi/strimzi-kafka-bridge/blob/master/src/main/java/io/strimzi/kafka/bridge/http/converter/HttpJsonMessageConverter.java) and [HttpBinaryMessageConverter](https://github.com/strimzi/strimzi-kafka-bridge/blob/master/src/main/java/io/strimzi/kafka/bridge/http/converter/HttpBinaryMessageConverter.java) testing both cases was a way to increase coverage.

see-quick replied:
Overall, if currently there are two different classes HttpJsonMessageConverter, HttpBinaryMessageConverter with the related logic now it makes sense. I do not know how were tests before but now it looks good.

scholzj commented:
@ppatierno What do we need to do here to move this PR forward?

ppatierno commented:
@scholzj we started a discussion about how to bring headers (which kind of data/format) here: https://github.com/strimzi/strimzi-kafka-bridge/pull/413#discussion_r429988641
I was hoping for feedback from you and @stanlyDoge to make a final decision.

ppatierno commented:
@codeJack can you fix the conflicts, please?

scholzj commented:
@ppatierno I don't have any strong opinions. But I guess it should follow the same format rules as the payload or key which can also have different formats in a single message. So I guess right now we support either both in JSON format or both in _binary_ (base64 encoded form). So similarly we should handle the headers?

ppatierno commented:
I was making the point that native Kafka clients don't make any assumption about headers values, they are just bytes. With payload, even if it's really just bytes as well, you have serialize/deserialize on top of it so somehow you have a way to interpret your data in a specific format.
I have no strong opinion as well, for this reason I was raising the discussion.
If we decide to support JSON/binary do you see it defined for each header (so the client can mix header JSON values with binary values) with `format` field or we should be just tied to the "format" specified when the consumer is created (which is related to the payload).
Anyway, let's remember that for the @codeJack use case, bringing just strings in the header values means JSON strings so double quotes has to be part of the value to be a valid JSON string.

codeJack commented:
@scholzj @ppatierno 

I can fix the conflicts, but I'd also like to align the PR to the final API design and I'm not sure we have reached an agreement yet.

Personally I don't have a strong opinion, both proposals have their pros and cons and for my own personal use-case - where I would mostly have _String_ based headers - it would be acceptable to base64-encode them upfront if we go for the **binary only** solution.

To summarize
- **binary only** would stick to the official APIs, offering the same level of service while being slightly less user friendly
- **String/binary** would be easier to use but, header-wise, it would not fully adhere to the currently existing Kafka APIs

If we go for the **String/Binary** solution I think we should specify the `format` at header level; if we define a single `format` that applies to all headers, as soon as you have a _binary_ one, you would need to to _base64_-encode all of your _String_ headers. Having to do that would kind of defeat the purpose of having a flexible - _String_ and _binary_ compatible - interface.

scholzj commented:
@ppatierno Any thoughts on this?

ppatierno commented:
> But I guess it should follow the same format rules as the payload or key which can also have different formats in a single message.

Actually it's not possible. When you create a consumer you specify the embedded format as binary or json. From that point in time, the consumer can get messages with key and value all binary or all json.
If we want to follow the same, then I think that for the headers we don't need to specify a format at all. We can just use the one on consumer creation. In this way, key/value in header follows the same rule as key/value in the payload.

@scholzj @codeJack does it sound reasonable to you?

scholzj commented:
Expecting the headers to be always in JSON seems a bit weird. The Kafka Java clients seem to handle this as String key and byte[] value. I wonder if we should somehow follow this although in reality it would be quite often that the value would be some String, Integer etc.

ppatierno commented:
@scholzj so you are saying what I raised few comments before? https://github.com/strimzi/strimzi-kafka-bridge/pull/413#discussion_r432604476

scholzj commented:
Yeah, looks like it :-). I guess handling the headers always as base64 encoded byte[] is the simple option ... but it is obviously user unfriendly. I wonder if we can have some autodetection of common types and handle all others as Base64. But no idea how easy / complicated that is and what would be the performance cost when you do it _per message_.

ppatierno commented:
@codeJack any more thoughts on this about having the headers as always raw bytes (as per Kafka native behavior) but of course Base64 encoded? I know that for taking just  a String is not the best solution (you have to encode your string in Base64) but I think we reflect the real behavior from the native Kafka clients. Wdyt?

scholzj reviewed (commented):
LGTM. Thanks for the PR.

ppatierno commented on src/main/resources/openapi.json:
```diff
@@ -1273,13 +1273,62 @@
+                    "value",
+                    "key"
+
+
```
can we remove these extra lines?

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/clients/Producer.java:
```diff
@@ -148,4 +132,44 @@ private Properties fillDefaultProperties() {
         return properties;
     }
+
+    public static class Builder {
```
I would name it `ProducerBuilder`

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/clients/Producer.java:
```diff
@@ -148,4 +132,44 @@ private Properties fillDefaultProperties() {
+        public Builder headers(List<KafkaHeader> headers) {
+            this.headers = headers;
+            return this;
+        }
```
should `headers` and `properties` prefixed with `with` (as for the null key record)?

ppatierno reviewed (commented):
@codeJack I had another pass.
@Frawless would you mind to take a look at the tests' part, please? Maybe you have a different opinion on the builder pattern added here.

Frawless reviewed (commented):
@ppatierno I agree with your suggestions

codeJack commented:
@ppatierno 
Can't reply directly to your comment about adding a description for the value field, so I will do it here :
I have also added the `"format": "byte"` qualifier as per [openapi specifications](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.3.md#data-types): whenever coupled with `"type":"string"` it stands for _base64 encoded characters_.

Also fixed the other points!

ppatierno reviewed (approved):
LGTM! Thanks for the PR!

ppatierno commented:
@codeJack let's wait for others approval ;-)

stanlyDoge reviewed (approved):
Thanks for your contribution. LGTM

stanlyDoge commented:
Please regenerate the doc to make build passing. 
```
Run the following to add up-to-date resources:
  make docu_api \
    && git add documentation/book/ \
    && git commit -s -m 'Update generated documentation'
```
